File: flutter-app/lib/utils/message_content.dart
enum MessageSegmentType { text, link, youtube }

class MessageSegment {
  const MessageSegment({
    required this.type,
    required this.value,
    this.videoId,
  });

  final MessageSegmentType type;
  final String value;
  final String? videoId;
}

final _urlRegex = RegExp(
  r'(https?:\/\/[^\s<]+[^\s<.,;:!?)\]}' "'" r'"])',
  caseSensitive: false,
);

final _youtubeRegex = RegExp(
  r'(?:https?:\/\/)?(?:www\.)?(?:youtube\.com\/(?:watch\?v=|embed\/|shorts\/)|youtu\.be\/)([a-zA-Z0-9_-]{11})',
  caseSensitive: false,
);

String? extractYouTubeId(String url) {
  final match = _youtubeRegex.firstMatch(url);
  return match?.group(1);
}

List<MessageSegment> splitMessageSegments(String text) {
  if (text.isEmpty) return [];

  final segments = <MessageSegment>[];
  var lastIndex = 0;

  for (final match in _urlRegex.allMatches(text)) {
    if (match.start > lastIndex) {
      segments.add(MessageSegment(
        type: MessageSegmentType.text,
        value: text.substring(lastIndex, match.start),
      ));
    }
    final url = match.group(0)!;
    final videoId = extractYouTubeId(url);
    segments.add(MessageSegment(
      type: videoId != null ? MessageSegmentType.youtube : MessageSegmentType.link,
      value: url,
      videoId: videoId,
    ));
    lastIndex = match.end;
  }

  if (lastIndex < text.length) {
    segments.add(MessageSegment(
      type: MessageSegmentType.text,
      value: text.substring(lastIndex),
    ));
  }

  return segments.isEmpty
      ? [MessageSegment(type: MessageSegmentType.text, value: text)]
      : segments;
}

const commonEmojis = [
  '😀', '😂', '😊', '😍', '🥰', '😎', '🤔', '😅', '🙏', '👍',
  '👏', '💪', '🔥', '✅', '❤️', '🎉', '⭐', '💯', '🏋️', '🥗',
  '📅', '📷', '💬', '👋', '🙂', '😢', '😡', '🤝', '✨', '🚀',
];

File: gym-admin/src/utils/messageContent.js
const URL_REGEX = /(https?:\/\/[^\s<]+[^\s<.,;:!?)\]}'"])/gi;
const YOUTUBE_REGEX = /(?:https?:\/\/)?(?:www\.)?(?:youtube\.com\/(?:watch\?v=|embed\/|shorts\/)|youtu\.be\/)([a-zA-Z0-9_-]{11})/i;

export function extractYouTubeId(url) {
  const match = String(url).match(YOUTUBE_REGEX);
  return match?.[1] || null;
}

export function splitMessageSegments(text) {
  const input = String(text || '');
  if (!input) return [];

  const segments = [];
  let lastIndex = 0;
  const regex = new RegExp(URL_REGEX.source, 'gi');
  let match;

  while ((match = regex.exec(input)) !== null) {
    if (match.index > lastIndex) {
      segments.push({ type: 'text', value: input.slice(lastIndex, match.index) });
    }
    const url = match[0];
    const videoId = extractYouTubeId(url);
    segments.push(videoId ? { type: 'youtube', value: url, videoId } : { type: 'link', value: url });
    lastIndex = match.index + url.length;
  }

  if (lastIndex < input.length) {
    segments.push({ type: 'text', value: input.slice(lastIndex) });
  }

  return segments.length ? segments : [{ type: 'text', value: input }];
}

export const COMMON_EMOJIS = [
  '😀', '😂', '😊', '😍', '🥰', '😎', '🤔', '😅', '🙏', '👍',
  '👏', '💪', '🔥', '✅', '❤️', '🎉', '⭐', '💯', '🏋️', '🥗',
  '📅', '📷', '💬', '👋', '🙂', '😢', '😡', '🤝', '✨', '🚀',
];
